### PR TITLE
Treat pointer literals as link-time constants

### PR DESCRIPTION
**Pull request by dougalm**
This allows standalone functions to close over array data.

### CODE REVIEW

File: src/lib/TopLevel.hs [hunk 1]
```diff
@@ -693,17 +696,17 @@ evalLLVM :: (Topper m, Mut n) => IxDestBlock n -> m n (Atom n)
 evalLLVM block = do
   backend <- backendName <$> getConfig
   logger  <- getFilteredLogger
-  (blockAbs, ptrVals) <- abstractPtrLiterals block
   let (cc, _needsSync) = case backend of LLVMCUDA -> (EntryFun CUDARequired   , True )
                                          _        -> (EntryFun CUDANotRequired, False)
   ImpFunctionWithRecon impFun reconAtom <- checkPass ImpPass $
-                                             toImpFunction backend cc blockAbs
+    toImpFunction backend cc $ Abs Empty block
```
Comment (apaszke) [inline src/lib/TopLevel.hs:702]: Perhaps remove the `Abs` too, since it's always empty?

**Review by apaszke — APPROVED** [review]
Nice! LGTM, but it would be nice to get rid of the pointer abs in `toImpFunction`. I don't think we use it

File: src/lib/Util.hs [hunk 2]
```diff
@@ -141,6 +141,15 @@ mapMaybe f (x:xs) = let rest = mapMaybe f xs
...
                         Nothing -> rest
 
+forMFilter :: Monad m => [a] -> (a -> m (Maybe b)) -> m [b]
+forMFilter [] _ = return []
+forMFilter (x:xs) f = do
+  ans <- f x
+  rest <- forMFilter xs f
+  return case ans of
+    Just y  -> y : rest
+    Nothing -> rest
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/Util.hs:151]: Good idea

File: src/lib/TopLevel.hs (hunk 1 above)
Comment (dougalm) [reply src/lib/TopLevel.hs:702]: Yes, good point. I had thought that this was used in the standalone function path too but I was wrong.

**Pull request merged**